File: Codes/0684-redundant-connection_tuoPuPaiXu.java
/*
 * @Author: LetMeFly
 * @Date: 2024-11-02 21:53:59
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2024-11-02 21:57:54
 */
class Solution {
    private int[] fa;

    private int find(int x) {
        if (fa[x] != x) {
            fa[x] = find(fa[x]);
        }
        return fa[x];
    }

    private void union(int x, int y) {
        fa[find(x)] = find(y);
    }

    public int[] findRedundantConnection(int[][] edges) {
        fa = new int[edges.length + 1];
        for (int i = 1; i <= edges.length; i++) {
            fa[i] = i;
        }
        for (int[] edge : edges) {
            if (find(edge[0]) == find(edge[1])) {
                return edge;
            } else {
                union(edge[0], edge[1]);
            }
        }
        return new int[0];
    }
}

File: Codes/0684-redundant-connection_tuoPuPaiXu.cpp
/*
 * @Author: LetMeFly
 * @Date: 2024-11-02 21:33:47
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2024-11-02 21:44:51
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
private:
    vector<int> fa;

    int find(int x) {
        if (fa[x] != x) {
            fa[x] = find(fa[x]);
        }
        return fa[x];
    }

    void union_(int x, int y) {
        fa[find(x)] = find(y);
    }
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        fa.resize(edges.size() + 1);
        for (int i = 1; i <= edges.size(); i++) {
            fa[i] = i;
        }
        // for (vector<int>& edge : edges) {
        //     union_(edge[0], edge[1]);
        // }
        // for (int i = edges.size() - 1; i > 0; i--) {
        //     if (find(edges[i][0]) == find(edges[i][1])) {
        //         return edges[i];
        //     }
        // }
        for (vector<int>& edge : edges) {
            if (find(edge[0]) == find(edge[1])) {
                return edge;
            } else {
                union_(edge[0], edge[1]);
            }
        }
        return {};  // FAKE RETURN
    }
};